internal/service/managedobjectstorage/managed_object_storage.go

```diff
@@ -0,0 +1,360 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func ResourceManagedObjectStorage() *schema.Resource {
+	return &schema.Resource{
+		Description:   "This resource represents an UpCloud Managed Object Storage instance, which provides S3 compatible storage.",
+		CreateContext: resourceManagedObjectStorageCreate,
+		ReadContext:   resourceManagedObjectStorageRead,
+		UpdateContext: resourceManagedObjectStorageUpdate,
+		DeleteContext: resourceManagedObjectStorageDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"configured_status": {
+				Description: "Service status managed by the end user.",
+				Required:    true,
+				Type:        schema.TypeString,
+				ValidateFunc: validation.StringInSlice(
+					[]string{
+						string(upcloud.ManagedObjectStorageConfiguredStatusStarted),
+						string(upcloud.ManagedObjectStorageConfiguredStatusStopped),
+					},
+					false,
+				),
+			},
+			"created_at": {
+				Description: "Creation time.",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+			"endpoint": {
+				Description: "Endpoints for accessing the Managed Object Storage service.",
+				Computed:    true,
+				Type:        schema.TypeSet,
+				Elem:        schemaEndpoint(),
+			},
+			"labels": utils.LabelsSchema("network gateway"),
+			"network": {
+				Description: "Attached networks from where object storage can be used. Private networks must reside in object storage region.",
+				Optional:    true,
+				Type:        schema.TypeSet,
+				Elem:        schemaNetwork(),
+			},
+			"operational_state": {
+				Description: "Operational state of the Managed Object Storage service.",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+			"region": {
+				Description: "Region in which the service will be hosted, see `upcloud_object_storage2_regions` data source.",
+				Required:    true,
+				Type:        schema.TypeString,
+			},
+			"updated_at": {
+				Description: "Creation time.",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+			"users": {
+				Description: "List of cloud Users allowed to use object storage. See `upcloud_managed_object_storage_user_access_key` for managing access keys.",
+				Optional:    true,
+				Type:        schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+func schemaEndpoint() *schema.Resource {
+	return &schema.Resource{
+		Description: "OpenSearch index",
+		Schema: map[string]*schema.Schema{
+			"domain_name": {
+				Description: "Domain name of the endpoint.",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+			"type": {
+				Description: "Type of the endpoint (`private` / `public`).",
+				Computed:    true,
+				Type:        schema.TypeString,
```

Comment:
Could add validation here, e.g.: 

```suggestion
				Type:        schema.TypeString,
				ValidateFunc: validation.StringInSlice([]string{"private", "public"}, false),
```

Reply:
Added, though not strictly necessary for a computed field.

Reply:
Ah, sorry, didn't notice the `Computed: true`, seems that `ValidateFunc` cannot be used in this case

Reply:
No worries, it was useless to add it in any case. :) Reverted the change just now.